File: problem/Longest_Substring_No_Repeat/Longest_Substring_No_Repeat.cpp
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> charIndex;
        int maxLen = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++)
        {
            if (charIndex.find(s[right]) != charIndex.end() &&
                charIndex[s[right]] >= left)
            {
                left = charIndex[s[right]] + 1;
            }

            charIndex[s[right]] = right;
            maxLen = max(maxLen, right - left + 1);
        }

        return maxLen;
    }
};

File: problem/Longest_Substring_No_Repeat/Longest_Substring_No_Repeat.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const charIndex = new Map();
    let maxLen = 0;
    let left = 0;
    
    for (let right = 0; right < s.length; right++) {
        if (charIndex.has(s[right]) && charIndex.get(s[right]) >= left) {
            left = charIndex.get(s[right]) + 1;
        }
        
        charIndex.set(s[right], right);
        maxLen = Math.max(maxLen, right - left + 1);
    }
    
    return maxLen;
};
